lib/mayaUsd/ufe/UsdUndoMaterialCommands.cpp

```diff
@@ -34,6 +35,45 @@ PXR_NAMESPACE_USING_DIRECTIVE
 namespace MAYAUSD_NS_DEF {
 namespace ufe {
 
+namespace {
+// We will not use the value of UsdUtilsGetMaterialsScopeName() for the material scope.
+static const std::string kDefaultMaterialScopeName("mtl");
+
+#if (UFE_PREVIEW_VERSION_NUM >= 4010)
+bool connectShaderToMaterial(
+    Ufe::SceneItem::Ptr shaderItem,
+    UsdPrim             materialPrim,
+    const std::string&  nodeId)
+{
+    auto                 shaderUsdItem = std::dynamic_pointer_cast<UsdSceneItem>(shaderItem);
+    auto                 shaderPrim = UsdShadeShader(shaderUsdItem->prim());
+    UsdShadeOutput       materialOutput;
+    PXR_NS::SdrRegistry& registry = PXR_NS::SdrRegistry::GetInstance();
+    PXR_NS::SdrShaderNodeConstPtr shaderNodeDef
+        = registry.GetShaderNodeByIdentifier(TfToken(nodeId));
+    SdrShaderPropertyConstPtr shaderOutputDef;
+    if (shaderNodeDef->GetSourceType() == "glslfx") {
+        materialOutput = UsdShadeMaterial(materialPrim).CreateSurfaceOutput();
+        shaderOutputDef = shaderNodeDef->GetShaderOutput(TfToken("surface"));
+    } else {
+        if (shaderNodeDef->GetOutputNames().size() != 1) {
+            TF_RUNTIME_ERROR(
+                "Cannot resolve which output of shader %s should be connected to surface",
+                nodeId.c_str());
+            return false;
+        }
+        materialOutput
+            = UsdShadeMaterial(materialPrim).CreateSurfaceOutput(shaderNodeDef->GetSourceType());
+        shaderOutputDef = shaderNodeDef->GetShaderOutput(shaderNodeDef->GetOutputNames()[0]);
+    }
+    UsdShadeOutput shaderOutput = shaderPrim.CreateOutput(
+        shaderOutputDef->GetName(), shaderOutputDef->GetTypeAsSdfType().first);
```

Comment:
Same question: could GetShaderOutput ever return a null pointer? This one seems safer, I assume the API guarantees returned output node are not null and are merely returned by pointer to allow sharing.